incubator/virtualcluster/example/cr-syncer/pkg/syncer/resource/project/controller.go

```diff
@@ -0,0 +1,290 @@
+}
+
+func NewProjectController(config *config.SyncerConfiguration,
+	client clientset.Interface,
+	informer informers.SharedInformerFactory,
+	vcClient vcclient.Interface,
+	vcInformer vcinformers.VirtualClusterInformer,
+	options manager.ResourceSyncerOptions) (manager.ResourceSyncer, error) {
+
+	klog.V(0).Info("start project controller")
+	err := createSuperProjectCRD(config)
+	if err !=nil {
+		return nil, fmt.Errorf("failed to create project crd: %v", err)
+	}
+
+	sc, err := dclient.New(config.RestConfig, dclient.Options{})
+	if err != nil {
+		klog.Errorf("failed to create super cluster client")
+	}
+
+	cSet := sets.NewString()
+	c := &controller{
+		BaseResourceSyncer: manager.BaseResourceSyncer{
+			Config: config,
+		},
+		config: config,
+		SuperClient: client.CoreV1(),
+		RestConfig:  config.RestConfig,
+		superClient: sc,
+		options:     options,
+		clusterSet:  cSet,
+	}
+	utilscheme.Scheme.AddKnownTypePair(&projectalpha1.Project{}, &projectalpha1.ProjectList{})
+
+	multiClusterProjectController, err := mc.NewMCController(&projectalpha1.Project{}, &projectalpha1.ProjectList{}, c,
+		mc.WithMaxConcurrentReconciles(constants.DwsControllerWorkerLow), mc.WithOptions(options.MCOptions))
+	if err != nil {
+		return nil, fmt.Errorf("failed to create project mc controller: %v", err)
+	}
+
+	c.multiClusterProjectController = multiClusterProjectController
+
+	projectList := &projectalpha1.ProjectList{}
+	err = c.superClient.List(context.Background(), projectList)
+	if err != nil {
+		klog.Errorf("cannot get cr project list")
+	}
+
+	c.projectcache, err = rinformer.New(config.RestConfig, rinformer.Options{})
+	if err != nil {
+		klog.Errorf("cannot get cr project cache %v", err)
+	}
+
+	projectInformer, err := c.projectcache.GetInformer(context.Background(), &projectalpha1.Project{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to create cr project informer: %v", err)
+	}
+
+	crdinformer , err := c.projectcache.GetInformer(context.Background(), &v1beta1.CustomResourceDefinition{})
+	if err != nil {
+		return nil, err
+	}
+
+	c.informer = projectInformer
+
+	if options.IsFake {
+		c.projectSynced = func() bool { return true }
+	} else {
+		c.projectSynced = projectInformer.HasSynced
+		c.vcSynced = vcInformer.Informer().HasSynced
+		c.crdSynced = crdinformer.HasSynced
+	}
+
+	projectPatroller, err := pa.NewPatroller(&projectalpha1.Project{}, c, pa.WithOptions(options.PatrolOptions))
+	if err != nil {
+		return nil, fmt.Errorf("failed to create project patroller: %v", err)
+	}
+	c.projectPatroller = projectPatroller
+
+	klog.Infof("new custom resource project creation done")
+	return c, nil
+}
+
+func createSuperProjectCRD(config *config.SyncerConfiguration) error {
+	apiextensionsClient, err := apiextensionsclientset.NewForConfig(config.RestConfig)
+	if err != nil {
+		return err
+	}
+
+	_, err = apiextensionsClient.ApiextensionsV1beta1().CustomResourceDefinitions().Get(context.TODO(), "projects.example.example.com", metav1.GetOptions{})
```

Comment:
Shouldn't this be done by the CRD syncer and having a CRD with the `super.public` annotation from the previous PR?

Reply:
Yes, this is just to add a Skeleton CRD in super cluster that is required by this CR syncer.. Without this, corresponding CR controller can not be initialized properly.
In most case, the CRD should have been deployed in super cluster. (as user decide to enable this CR syncer). Therefore we just do a get first and deploy the CRD in case it is not found.

Reply:
I would remove this and put this as a part of the docs to `kubectl apply -f` a CRD file into the persons local cluster when trying to learn from this example. Having this in here could confuse someone into thinking the syncer needed to manage the CRDs for the super cluster too